### PR TITLE
KAFKA-5975: No response when deleting topics and delete.topic.enable=false

### PR DESCRIPTION
**Pull request by mmolimar**
This PR implements [KIP-322](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=87295558).

### CODE REVIEW

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 1]
```diff
@@ -1394,9 +1394,10 @@ class KafkaApis(val requestChannel: RequestChannel,
       sendResponseMaybeThrottle(request, createResponse)
     }
 
-    if (!controller.isActive) {
+    if (!controller.isActive || !config.deleteTopicEnable) {
+      val error = if (!controller.isActive) Errors.NOT_CONTROLLER else Errors.NONE
```
Comment (tombentley) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: It can't be correct that the caller is left with the impression that the deletion completed successfully when, in fact, deletion is disabled and the topic still exists. Admittedly, none of the other existing errors looks like a good fit either. Perhaps there needs to be a new `Error.INCOMPATIBLE_WITH_CONFIG` or something for cases where the configuration prevents fulfilment of the request.

Comment (mmolimar) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I thought the same thing but I didn't include that error because, for example, the AdminManager swallows an exception in case the topic is already marked for deletion or even the AdminUtils writes in Zk the topic as deleted but the TopicDeletionManager will never do.

Anyway, I just added that new error with its corresponding exception :-)

**Comment by mmolimar** [comment]
Please review any of @tombentley @miguno @ewencp @guozhangwang @dguy @mjsax

File: clients/src/main/java/org/apache/kafka/common/errors/IncompatibleWithConfigException.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.common.errors;
+
+public class IncompatibleWithConfigException extends ApiException {
```
Comment (tombentley) [inline clients/src/main/java/org/apache/kafka/common/errors/IncompatibleWithConfigException.java]: It would be helpful to add Javadoc explaining what this exception is for, and how it differs from, for example, `PolicyViolationException`.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java [hunk 3]
```diff
@@ -97,7 +98,7 @@
  * Do not add exceptions that occur only on the client or only on the server here.
  */
 public enum Errors {
-    UNKNOWN_SERVER_ERROR(-1, "The server experienced an unexpected error when processing the request",
+    UNKNOWN_SERVER_ERROR(-1, "The server experienced an unexpected error when processing the request.",
```
Comment (tombentley) [inline clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Since these exception messages are unrelated to the bug you're fixing, it would be better to leave the inconsistent punctuation for another PR.

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java [hunk 4]
```diff
@@ -538,13 +539,20 @@ public ApiException build(String message) {
...
-    REASSIGNMENT_IN_PROGRESS(60, "A partition reassignment is in progress",
+    REASSIGNMENT_IN_PROGRESS(60, "A partition reassignment is in progress.",
         new ApiExceptionBuilder() {
             @Override
             public ApiException build(String message) {
                 return new ReassignmentInProgressException(message);
             }
-    });
+    }),
+    INCOMPATIBLE_WITH_CONFIG(61, "The operation requested is incompatible with the configuration.",
```
Comment (tombentley) [inline clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: "incompatible with the *broker's* configuration" would be a bit clearer, I think.

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 5]
```diff
@@ -1402,9 +1402,10 @@ class KafkaApis(val requestChannel: RequestChannel,
       sendResponseMaybeThrottle(request, createResponse)
     }
 
-    if (!controller.isActive) {
+    if (!controller.isActive || !config.deleteTopicEnable) {
+      val error = if (!controller.isActive) Errors.NOT_CONTROLLER else Errors.INCOMPATIBLE_WITH_CONFIG
```
Comment (tombentley) [inline core/src/main/scala/kafka/server/KafkaApis.scala]: I don't think we can return a whole new error code without an explicit change in the binary format (think about what happens in a 11.0.0 client gets this new error code from a 11.0.2 broker, for example). For old clients we could use an existing error code. `POLICY_VIOLATION` has similar semantics, so maybe that would be a good choice, though @ijuma would be better placed to advise about this.

Comment (tombentley) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Though thinking about it, this API doesn't currently return `POLICY_VIOLATION`, so even that would probably need a new binary version (@ijuma?). In which case it may be better to fix in as part of KIP-201.

Comment (ijuma) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Yes, we should try to reuse an error code that is already used by this protocol API.

Comment (mmolimar) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: Which one do you think fits better? ``INVALID_REQUEST``?

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 3 above)
Comment (mmolimar) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: Just for the shake of this PR, I think we could maintain these very small changes...

File: clients/src/main/java/org/apache/kafka/common/protocol/Errors.java (hunk 4 above)
Comment (mmolimar) [reply clients/src/main/java/org/apache/kafka/common/protocol/Errors.java]: I removed this error code.

File: core/src/main/scala/kafka/server/KafkaApis.scala (hunk 5 above)
Comment (mmolimar) [reply core/src/main/scala/kafka/server/KafkaApis.scala]: I used ``INVALID_REQUEST``, ok?

**Comment by omkreddy** [comment]
@mmolimar  I have created KIP: https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=87295558 to return new error code for DeleteTopics API when topic deletion disabled.

Since you raised the PR for this, would you like to implement the proposed changes in your PR?
Let me know If you are busy, I can implement the changes on top of your PR.

**Comment by mmolimar** [comment]
Sure @omkreddy! I can do that.
But what about [this](https://github.com/apache/kafka/pull/5130) PR? Do I merge those changes into my PR?

**Comment by omkreddy** [comment]
@mmolimar Yes, you can merge my PR. Thanks.

**Comment by mmolimar** [comment]
Great. Will do

File: core/src/main/scala/kafka/server/KafkaApis.scala [hunk 6]
```diff
@@ -1538,6 +1538,12 @@ class KafkaApis(val requestChannel: RequestChannel,
         (topic, Errors.NOT_CONTROLLER)
       }.toMap
       sendResponseCallback(results)
+    } else if (!config.deleteTopicEnable) {
+      val error = if (request.context.apiVersion < 3) Errors.INVALID_REQUEST else Errors.TOPIC_DELETION_DISABLED
```
Comment (omkreddy) [inline core/src/main/scala/kafka/server/KafkaApis.scala:1543]: We also need to update DeleteTopicsRequest/DeleteTopicsResponse request/response classes to V3.

Comment (mmolimar) [reply core/src/main/scala/kafka/server/KafkaApis.scala:1543]: Done!

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsRequest.java [hunk 7]
```diff
@@ -51,9 +51,11 @@
      */
     private static final Schema DELETE_TOPICS_REQUEST_V2 = DELETE_TOPICS_REQUEST_V1;
 
+    private static final Schema DELETE_TOPICS_REQUEST_V3 = DELETE_TOPICS_REQUEST_V2;
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsRequest.java]: Can we add a comment like above?

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java [hunk 8]
```diff
@@ -53,8 +53,11 @@
      */
     private static final Schema DELETE_TOPICS_RESPONSE_V2 = DELETE_TOPICS_RESPONSE_V1;
 
+    private static final Schema DELETE_TOPICS_RESPONSE_V3 = DELETE_TOPICS_RESPONSE_V2;
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java]: Can we add a comment like above?

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsRequest.java (hunk 7 above)
Comment (omkreddy) [reply clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsRequest.java]: We need to include version 3 to switch case at line no: 121

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java [hunk 9]
```diff
@@ -53,8 +53,11 @@
...
+
     public static Schema[] schemaVersions() {
-        return new Schema[]{DELETE_TOPICS_RESPONSE_V0, DELETE_TOPICS_RESPONSE_V1, DELETE_TOPICS_RESPONSE_V2};
+        return new Schema[]{DELETE_TOPICS_RESPONSE_V0, DELETE_TOPICS_RESPONSE_V1,
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java:63]: check style error at line no: 60. 

`./gradlew :clients:checkstyleMain`

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  @Test
+  def testDeleteRecordsRequest() {
+
+    val topic = "topic-1"
+    val request = new DeleteTopicsRequest.Builder(Set(topic).asJava, 1000).build();
+    val response = sendDeleteTopicsRequest(request)
+
+    assertEquals(Errors.INVALID_REQUEST, response.errors.get(topic))
```
Comment (omkreddy) [inline core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala]: This should be TOPIC_DELETION_DISABLED. 

We can also add test for v2 requests 

```
val v2request = new DeleteTopicsRequest.Builder(Set(topic).asJava, 1000).build(2)
val v2response = sendDeleteTopicsRequest(v2request)
assertEquals(Errors.INVALID_REQUEST, v2response.errors.get(topic))
```

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java [hunk 11]
```diff
@@ -53,8 +53,14 @@
...
+    /**
+     * v3 request is the same that as v2. The response is different based on the request version.
+     */
+    private static final Schema DELETE_TOPICS_RESPONSE_V3 = DELETE_TOPICS_RESPONSE_V2;
+
     public static Schema[] schemaVersions() {
-        return new Schema[]{DELETE_TOPICS_RESPONSE_V0, DELETE_TOPICS_RESPONSE_V1, DELETE_TOPICS_RESPONSE_V2};
+        return new Schema[]{DELETE_TOPICS_RESPONSE_V0, DELETE_TOPICS_RESPONSE_V1,
+            DELETE_TOPICS_RESPONSE_V2, DELETE_TOPICS_RESPONSE_V3};
     }
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java:67]: Can we add the new error codes (TOPIC_DELETION_DISABLED, INVALID_REQUEST)  to the below comment?

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java [hunk 12]
```diff
@@ -53,8 +53,14 @@
      */
     private static final Schema DELETE_TOPICS_RESPONSE_V2 = DELETE_TOPICS_RESPONSE_V1;
 
+    /**
+     * v3 request is the same that as v2. The response is different based on the request version.
```
Comment (omkreddy) [inline clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java:57]: Could you add more specific description of the change? 
 something like  "v3 response can return TOPIC_DELETION_DISABLED exception"

**Review by omkreddy — COMMENTED** [review]
LGTM.

**Comment by omkreddy** [comment]
@hachikuji  Please take a look.

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java (hunk 12 above)
Comment (mmolimar) [reply clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java:57]: Done!

File: clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java (hunk 11 above)
Comment (mmolimar) [reply clients/src/main/java/org/apache/kafka/common/requests/DeleteTopicsResponse.java:67]: Done!

**Comment by omkreddy** [comment]
@lindong28  @harshach  Can you take a look at this PR?

**Comment by lindong28** [comment]
Sure. I will review it later this week.

**Comment by harshach** [comment]
@omkreddy will take a look this week.

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+import kafka.network.SocketServer
+import kafka.utils._
+import org.apache.kafka.common.protocol.{ApiKeys, Errors}
+import org.apache.kafka.common.requests.{DeleteTopicsRequest, DeleteTopicsResponse}
+import org.junit.Assert._
+import org.junit.Test
+
+import scala.collection.JavaConverters._
+
+class DeleteTopicsRequestWithDeletionDisabledTest extends BaseRequestTest {
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala:29]: Should we have an integration test for AdminClient as well?

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+      enableControlledShutdown = false, enableDeleteTopic = false,
+      interBrokerSecurityProtocol = Some(securityProtocol),
+      trustStoreFile = trustStoreFile, saslProperties = serverSaslProperties, logDirCount = logDirCount)
+    props.foreach(propertyOverrides)
+    props.map(KafkaConfig.fromProps)
+  }
+
+  @Test
+  def testDeleteRecordsRequest() {
+
```
Comment (hachikuji) [inline core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala]: nit: unneeded newline

**Review by hachikuji — COMMENTED** [review]
Thanks, looks good. Left a couple minor comments.

**Comment by lindong28** [comment]
Ah I am sorry @omkreddy! I obviously forgot to review this patch.. Please ping me next time you have a patch and I will review it as soon as possible. Thanks!

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala (hunk 14 above)
Comment (mmolimar) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala]: Done!

File: core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala (hunk 13 above)
Comment (mmolimar) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala:29]: I can include another class overwriting the prop ``delete.topic.enable`` to test it. Not a big deal. I think it can be useful, what do you say?

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala:29]: I was thinking the benefit is that it verifies the error handling in the AdminClient. We could also do a simple unit test.

Comment (omkreddy) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala:29]: maybe we can add another test to this class. we can borrow some code from [AdminClientIntegrationTest.scala](https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala#L211)

Comment (mmolimar) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala:29]: That's what I was saying @omkreddy. It'd be write another class extending ``IntegrationTestHarness`` disabling the property and including this test.
But maybe adding a simple unit test in ``KafkaAdminClientTest``as @hachikuji suggested it'd be fine (btw, we don't have unit tests to remove topics in that class).
So?

Comment (hachikuji) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala:29]: @mmolimar I'm ok either way. I'd probably lean toward a unit test in `KafkaAdminClientTest` since integration tests are more expensive for the build time. And yeah, there does seem to be a testing gap for the error handling of this API.

Comment (mmolimar) [reply core/src/test/scala/unit/kafka/server/DeleteTopicsRequestWithDeletionDisabledTest.scala:29]: Added! ;-)

**Review by hachikuji — APPROVED** [review]
@mmolimar Thanks for the patch, LGTM. I'll merge once you've had a chance to rebase.

**Comment by mmolimar** [comment]
@hachikuji Conflicts resolved :-)
